internal/webhook/external/refresh.go

```diff
@@ -103,3 +125,61 @@ func refreshWarehouses(
 		},
 	)
 }
+
+func needsRefresh(
+	_ context.Context,
+	subs []kargoapi.RepoSubscription,
+	qualifier string,
+	repoURLs ...string,
+) (*bool, error) {
+	var shouldRefresh bool
+	subs = filterSubsByRepoURL(subs, repoURLs...) // only interested in subs that contain any of the repo URLs.
+	for _, s := range subs {
+		switch {
+		case s.Git != nil:
+			selector, err := commit.NewSelector(*s.Git, nil)
+			if err != nil {
+				return nil, fmt.Errorf("error creating commit selector for Git subscription %q: %w",
+					s.Git.RepoURL, err,
+				)
+			}
+			shouldRefresh = selector.MatchesRef(qualifier)
+		case s.Image != nil:
+			selector, err := image.NewSelector(*s.Image, nil)
+			if err != nil {
+				return nil, fmt.Errorf("error creating image selector for Image subscription %q: %w",
+					s.Image.RepoURL, err,
+				)
+			}
+			shouldRefresh = selector.MatchesTag(qualifier)
+		case s.Chart != nil:
+			selector, err := chart.NewSelector(*s.Chart, nil)
+			if err != nil {
+				return nil, fmt.Errorf("error creating chart selector for Chart subscription %q: %w",
+					s.Chart.RepoURL, err,
+				)
+			}
+			shouldRefresh = selector.MatchesVersion(qualifier)
+		}
+		if shouldRefresh {
+			// exit early if we already found a match
+			return &shouldRefresh, nil
+		}
+	}
+	return &shouldRefresh, nil
+}
+
+// filterSubsByRepoURL deletes all subscriptions from subs that do not
+// match any of the provided repository URLs; omitting them from processing.
+func filterSubsByRepoURL(subs []kargoapi.RepoSubscription, repoURLs ...string) []kargoapi.RepoSubscription {
+	containsRepoURL := func(sub kargoapi.RepoSubscription) bool {
+		return sub.Image != nil && slices.Contains(repoURLs, sub.Image.RepoURL) ||
+			sub.Git != nil && slices.Contains(repoURLs, sub.Git.RepoURL) ||
```

Comment:
You forgot to normalize `sub.Image.RepoURL` and `sub.Git.RepoURL`. (You remembered for `sub.Chart.RepoURL`.)

btw, we know, contextually, the way this is used, `repoURLs` will already all be normalized. We should probably document that expectation in the comment for this func.

Reply:
Also some extra parens here would make the logic easier to follow without needing to do any head scratching about the precedence of `&&` vs `||`. i.e.:

```
(sub.Image != nil && slices.Contains(...)) ||
(sub.Git != nil && slices.Contains(...)) ||
(sub.Chart != nil && slices.Contains(...))
```